### ISSUE TITLE
Variable compared to another variable

### ISSUE THREAD

**Issue by msscaroso** [venmo/business-rules#48]
Currently, a variable is compared to a value, then if the rule is triggered, the action is taken.
The `variable` comes from the system. It's possible to make a complex logic to compute those variables.

But `values`,  the ones that are defined during the rule creation, are purely static and simple. 
I believe allowing variables to be compared to variables would be very useful, allowing to created more generic rules.

One example: Suppose that we have `Users` that can consume a `Resource`, but each user have a consumption limitation. That limitation is specific to each user. Thus, in order to restrain the user consumption, we would need to create one business rule for each user, specifying the user capacity in each rule.
If it was possible to compare a variable to another variable, just one rule would be necessary by doing: `user.resources_count_variable` < `user.max_resource`. Then we could delegate the max_resource definition to another entity.

**Comment by bheuse** [comment]
Variable compared to another variable: I would appreciate this too ...

Syntax:
              "name": "current_month",
              "operator": "equal_to",
              "value": "$sales_month"
Would this code fix works ? Any side effect ? any other places where we need to consider changes ?

def check_condition(condition, defined_variables):
    """ Checks a single rule condition - the condition will be made up of
    variables, values, and the comparison operator. The defined_variables
    object must have a variable defined for any variables in this condition.
    """
   # Added Pseudo Code
    if condition['value'].startswith("$"):
      name = re.sub("^$","",condition['value'])
      condition['value']= _get_variable_value(defined_variables, name) 
   # End Pseudo Code
      

    name, op, value = condition['name'], condition['operator'], condition['value']
    operator_type = _get_variable_value(defined_variables, name)
    return _do_operator_comparison(operator_type, op, value)

**Comment by Ana-Htor** [comment]
Hi,

@bheuse I tried your proposition but it doesn't work as the _get_variable_value function do not return the value of the variable itself but the object : Numeric object for instance.

You should rather create a custom _get_variable_function, called for instance _get_variable_function_custom, and defined as such

```python
def _get_variable_value_custom(defined_variables, name):
    def fallback(*args, **kwargs):
        raise AssertionError("Variable {0} is not defined in class {1}".format(
                name, defined_variables.__class__.__name__))
    method = getattr(defined_variables, name, fallback)
    val = method()
    return val
```
The function is the same as _get_variable_value except that the return is set to val and not method

Then you can keep your proposition but calling the custom function in your if block:

```python
def check_condition(condition, defined_variables):
    name, op, value = condition['name'], condition['operator'], condition['value']

    # Begin Added Code
    if (str(value).startswith("$")):
        var_name = str(value).replace("$","")
        value = _get_variable_value_custom(defined_variables, var_name) #Here call the custom function instead of the old one
    # End Added Code

    operator_type = _get_variable_value(defined_variables, name)
    return _do_operator_comparison(operator_type, op, value)

``` 

It works for me. Hope will help.

Cheers.
